File: stringart.py
import numpy as np
from PIL import Image
from skimage.transform import radon
import cv2


def crop_image_to_circle(image_path, output_path):
    # Read the input image
    image = cv2.imread(image_path, cv2.IMREAD_UNCHANGED)

    # Ensure the image has four channels (with alpha)
    if image.shape[2] == 3:
        image = cv2.cvtColor(image, cv2.COLOR_BGR2BGRA)

    # Create a circular mask
    height, width, _ = image.shape
    center = (width // 2, height // 2)
    radius = min(center[0], center[1])
    mask = np.zeros((height, width), dtype=np.uint8)
    cv2.circle(mask, center, radius, 255, -1)

    # Apply the mask to the image
    result = cv2.bitwise_and(image, image, mask=mask)

    # Save the output image
    cv2.imwrite(output_path, result)


def radon_string(s_string, alpha_string, radius):
    proj_angle = np.linspace(0.0, 180.0, 512, endpoint=False)
    proj_pos = np.linspace(0.0, 512.0, 512, endpoint=False)
    radon_transform = np.zeros((512, 512))

    for alpha_idx, alpha in enumerate(proj_angle):
        for s_idx, s in enumerate(proj_pos):
            sin_angle = np.sin(np.deg2rad(alpha - alpha_string))
            if sin_angle == 0:
                radon_transform[alpha_idx, s_idx] = 0
                break

            evaluator = (
                s**2
                + s_string**2
                - 2 * s * s_string * np.cos(np.deg2rad(alpha - alpha_string))
            ) / (sin_angle**2)
            if evaluator > radius**2:
                radon_transform[alpha_idx, s_idx] = 0
            else:
                radon_transform[alpha_idx, s_idx] = 1 / np.abs(sin_angle)

    return radon_transform


def radon_lines():
    n = 300
    circle_theta1 = np.linspace(0.0, 360.0, n, endpoint=False)
    circle_theta2 = np.linspace(0.0, 360.0, n, endpoint=False)

    total_elements = (
        n * (n - 1)
    ) // 2  # Number of unique pairs (theta1, theta2) with theta1 < theta2
    line_theta_lookup = np.zeros((total_elements, 2))
    line_radon_lookup = np.zeros(total_elements)

    index = 0
    for i, theta1 in enumerate(circle_theta1):
        for j, theta2 in enumerate(circle_theta2):
            if theta1 >= theta2:
                continue
            else:
                radius = 256
                s_string = radius * np.cos(np.deg2rad(theta2 - theta1) / 2)
                alpha_string = (theta2 + theta1) / 2
                line_radon_lookup[index] = radon_string(s_string, alpha_string, radius)
                line_theta_lookup[index] = np.array([theta1, theta2])
                index += 1

    return line_radon_lookup, line_theta_lookup


def get_strings(img_path):  # using max
    # Load and convert the image to grayscale
    img = Image.open(img_path).convert("L")
    img_array = np.array(img)
    radius = 256
    theta = np.linspace(0.0, 180.0, max(img_array.shape), endpoint=False)

    # Perform Radon transform
    sinogram_img = radon(img_array, theta=theta, circle=True)
    len_adj_sinogram = np.zeros_like(sinogram_img)
    strings = []

    for _ in range(5):  # Extracting top strings
        for s in range(radius):
            len_of_proj = 2 * ((radius**2 - s**2) ** 0.5)
            len_adj_sinogram[s] = sinogram_img[s, :] / len_of_proj

        max_index = np.argmax(len_adj_sinogram)
        s_idx, theta_idx = max_index

        alpha = theta[theta_idx]
        s = s_idx

        theta1 = alpha - np.rad2deg(np.arccos(s / radius))
        theta2 = alpha + np.rad2deg(np.arccos(s / radius))
        print((theta1, theta2))
        strings.append((theta1, theta2))

        # Subtracting the identified string's radon transform
        sinogram_img = np.subtract(sinogram_img, radon_string(s, alpha, radius))

    return strings


def greedy_solve(img_path):
    radon_string = np.load("radon_lookup.npy")
    img = Image.open(img_path).convert("L")
    img_array = np.array(img)

    theta = np.linspace(0.0, 180.0, max(img_array.shape), endpoint=False)

    # Perform Radon transform
    radon_img = radon(img_array, theta=theta, circle=True)

    m, n, p = radon_string.shape  # A is 3D: m x n x p
    x = np.zeros(p)  # x is 1D of length p

    for i in range(p):
        x[i] = 1
        current_residual = sum(
            np.linalg.norm(np.dot(radon_string[:, :, k], x) - radon_img[:, k])
            for k in range(n)
        )

        x[i] = 0
        if current_residual < sum(
            np.linalg.norm(np.dot(radon_string[:, :, k], x) - radon_img[:, k])
            for k in range(n)
        ):
            x[i] = 1

    return x


def create_string_art(strings):
    canvas = np.zeros((512, 512), dtype=np.uint8)
    circle_center = [512 // 2, 512 // 2]
    radius = 256

    for string in strings:
        theta1 = np.deg2rad(string[0])
        theta2 = np.deg2rad(string[1])

        x1 = int(circle_center[0] + radius * np.cos(theta1))
        y1 = int(circle_center[1] + radius * np.sin(theta1))
        x2 = int(circle_center[0] + radius * np.cos(theta2))
        y2 = int(circle_center[1] + radius * np.sin(theta2))

        cv2.line(canvas, (x1, y1), (x2, y2), 255, 1)

    return canvas


# art = create_string_art([tuple(np.random.uniform(0, 360, size=2)) for _ in range(300)])
art = create_string_art(get_strings("circle.png"))

# Save and display the result
cv2.imwrite("art.png", art)
cv2.imshow("Canvas", art)
cv2.waitKey(0)
cv2.destroyAllWindows()

File: string_art.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <cmath>
#include <stdexcept>

using namespace cv;
using namespace std;

Mat crop_image_to_circle(const string &image_path, const string &output_path, int max_length = 3840)
{
    Mat image = imread(image_path, IMREAD_UNCHANGED);
    if (image.empty())
    {
        throw runtime_error("Could not open or find the image");
    }

    int height = image.rows;
    int width = image.cols;

    if (max(width, height) > max_length)
    {
        if (width > height)
        {
            int new_width = max_length;
            int new_height = static_cast<int>(height * max_length / width);
            resize(image, image, Size(new_width, new_height), 0, 0, INTER_AREA);
        }
        else
        {
            int new_height = max_length;
            int new_width = static_cast<int>(width * max_length / height);
            resize(image, image, Size(new_width, new_height), 0, 0, INTER_AREA);
        }
    }

    height = image.rows;
    width = image.cols;
    Point center(width / 2, height / 2);
    int radius = min(center.x, center.y);
    Mat mask = Mat::zeros(height, width, CV_8UC1);
    circle(mask, center, radius, Scalar(255), -1);

    Mat result;
    if (image.channels() == 1)
    {
        image.copyTo(result, mask);
    }
    else
    {
        Mat masked_image;
        vector<Mat> channels;
        split(image, channels);
        for (auto &channel : channels)
        {
            Mat masked_channel;
            channel.copyTo(masked_channel, mask);
            channels.push_back(masked_channel);
        }
        merge(channels, result);
    }

    int square_size = radius * 2;
    int crop_x = (width - square_size) / 2;
    int crop_y = (height - square_size) / 2;
    Mat cropped_square = result(Rect(crop_x, crop_y, square_size, square_size));

    Mat output_image = Mat::zeros(square_size, square_size, image.type());
    cropped_square.copyTo(output_image(Rect(0, 0, cropped_square.cols, cropped_square.rows)));

    imwrite(output_path, output_image);

    return output_image;
}

Mat radon_string(double s_string, double alpha_string, int radius)
{
    int size = radius * 2;
    Mat radon_transform = Mat::zeros(size, size, CV_64F);
    vector<double> proj_angle(size);
    vector<double> proj_pos(size);

    for (int i = 0; i < size; ++i)
    {
        proj_angle[i] = i * 180.0 / size;
        proj_pos[i] = i - radius;
    }

    Point max_index(0, 0);
    for (int alpha_idx = 0; alpha_idx < size; ++alpha_idx)
    {
        double alpha = proj_angle[alpha_idx];
        double sin_angle = sin((alpha - alpha_string) * CV_PI / 180.0);
        if (abs(sin_angle) == 0)
        {
            max_index.y = alpha_idx;
            continue;
        }

        for (int s_idx = 0; s_idx < size; ++s_idx)
        {
            double s = proj_pos[s_idx];
            if (s == s_string)
            {
                max_index.x = s_idx;
            }

            double evaluator = (s * s + s_string * s_string - 2 * s * s_string * cos((alpha - alpha_string) * CV_PI / 180.0)) / (sin_angle * sin_angle);
            if (evaluator > radius * radius)
            {
                radon_transform.at<double>(s_idx, alpha_idx) = 0;
            }
            else
            {
                radon_transform.at<double>(s_idx, alpha_idx) = 1 / abs(sin_angle);
            }
        }
    }

    radon_transform.at<double>(max_index.x, max_index.y) = radon_transform.at<double>(max_index.x, max_index.y) * 1.1;
    normalize(radon_transform, radon_transform, 0, 1, NORM_MINMAX);

    return radon_transform;
}

void plot_sinogram(const Mat &sinogram)
{
    Mat display;
    normalize(sinogram, display, 0, 255, NORM_MINMAX);
    display.convertTo(display, CV_8U);
    applyColorMap(display, display, COLORMAP_HOT);
    imshow("Sinogram", display);
    waitKey(0);
}

pair<vector<pair<double, double> >, int> get_strings(const string &img_path, int num_of_anchors)
{
    Mat image = imread(img_path, IMREAD_GRAYSCALE);
    if (image.empty())
    {
        throw runtime_error("Could not open or find the image");
    }

    image = 255 - image;

    int height = image.rows;
    int width = image.cols;
    Point center(width / 2, height / 2);
    int radius = min(center.x, center.y);

    vector<double> theta(radius * 2);
    for (int i = 0; i < radius * 2; ++i)
    {
        theta[i] = i * 180.0 / (radius * 2);
    }

    Mat sinogram_img = Mat::zeros(radius * 2, radius * 2, CV_64F);
    for (int i = 0; i < radius * 2; ++i)
    {
        for (int j = 0; j < radius * 2; ++j)
        {
            double angle = theta[j] * CV_PI / 180.0;
            double sum = 0;
            for (int k = 0; k < radius * 2; ++k)
            {
                int x = center.x + (k - radius) * cos(angle);
                int y = center.y + (k - radius) * sin(angle);
                if (x >= 0 && x < width && y >= 0 && y < height)
                {
                    sum += image.at<uchar>(y, x);
                }
            }
            sinogram_img.at<double>(i, j) = sum;
        }
    }

    normalize(sinogram_img, sinogram_img, 0, 1, NORM_MINMAX);

    vector<pair<double, double> > strings;
    Mat len_adj_sinogram = Mat::zeros(sinogram_img.size(), CV_64F);
    for (int n = 0; n < num_of_anchors; ++n)
    {
        for (int s = 0; s < radius * 2; ++s)
        {
            if (s == 0)
            {
                len_adj_sinogram.row(s).setTo(0);
            }
            else
            {
                double len_of_proj = 2 * sqrt(radius * radius - (s - radius) * (s - radius));
                len_adj_sinogram.row(s) = sinogram_img.row(s) / len_of_proj;
            }
        }

        Point max_index;
        minMaxLoc(len_adj_sinogram, nullptr, nullptr, nullptr, &max_index);

        int s_idx = max_index.y;
        int theta_idx = max_index.x;

        double alpha = theta[theta_idx];
        double s = s_idx - radius;

        double theta1 = alpha - acos(s / radius) * 180.0 / CV_PI;
        double theta2 = alpha + acos(s / radius) * 180.0 / CV_PI;

        strings.emplace_back(theta1, theta2);

        Mat sinogram_string = radon_string(s, alpha, radius);
        sinogram_img -= sinogram_string;
        sinogram_img.setTo(0, sinogram_img < 0);

        cout << "String " << strings.size() << ": " << theta1 << " to " << theta2 << endl;
    }

    plot_sinogram(sinogram_img);

    return {strings, radius};
}

Mat create_string_art(const vector<pair<double, double> > &strings, int radius)
{
    Mat canvas = Mat::zeros(radius * 2, radius * 2, CV_8UC1);
    Point circle_center(radius, radius);

    for (const auto &string : strings)
    {
        double theta1 = string.first * CV_PI / 180.0;
        double theta2 = string.second * CV_PI / 180.0;

        Point pt1(circle_center.x + radius * cos(theta1), circle_center.y + radius * sin(theta1));
        Point pt2(circle_center.x + radius * cos(theta2), circle_center.y + radius * sin(theta2));

        line(canvas, pt1, pt2, Scalar(255), 1);
    }

    return canvas;
}

int main()
{
    try
    {
        Mat cropped_image = crop_image_to_circle("art.png", "test.png");
        auto [strings, radius] = get_strings("test.png", 900);
        Mat art = create_string_art(strings, radius);
        imwrite("test_900.png", art);
    }
    catch (const exception &e)
    {
        cerr << "Error: " << e.what() << endl;
        return 1;
    }

    return 0;
}